onnx/defs/tensor/defs.cc

```diff
@@ -403,46 +403,75 @@ ONNX_OPERATOR_SET_SCHEMA(
-              axis += rank;
-            }
-            const auto& splitDim = shape.dim(axis);
-            if (!splitDim.has_dim_value()) {
-              return;
-            }
-            int splitDimValue = static_cast<int>(splitDim.dim_value());
-            int chunkSize =
-                splitDimValue / static_cast<int>(ctx.getNumOutputs());
-            int leftOver = splitDimValue -
-                (chunkSize * static_cast<int>(ctx.getNumOutputs()));
-            for (int i = 0; i < static_cast<int>(ctx.getNumOutputs()); i++) {
-              split.push_back(i < leftOver ? chunkSize + 1 : chunkSize);
-            }
 
+          const auto& shape = ctx.getInputType(0)->tensor_type().shape();
+          int rank = shape.dim_size();
+          int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
+          if (axis < -rank || axis >= rank) {
+            fail_type_inference(
+                "Invalid value of attribute 'axis'. Rank=",
+                rank,
+                " Value=",
+                axis);
+          }
+          if (axis < 0) {
+            axis += rank;
+          }
+          const auto& splitDim = shape.dim(axis);
+          if (!splitDim.has_dim_value()) {
             for (size_t i = 0; i < ctx.getNumOutputs(); i++) {
               *ctx.getOutputType(i)->mutable_tensor_type()->mutable_shape() =
                   shape;
               ctx.getOutputType(i)
                   ->mutable_tensor_type()
                   ->mutable_shape()
                   ->mutable_dim(axis)
-                  ->set_dim_value(split[i]);
+                  ->Clear();
             }
+            return;
+          }
+          int splitDimValue = static_cast<int>(splitDim.dim_value());
+
+          std::vector<int64_t> split;
+          if (getRepeatedAttribute(ctx, "split", split)) {
+            if (split.size() != ctx.getNumOutputs()) {
+              fail_shape_inference(
+                  "Mismatch between number of splits (",
+                  split.size(),
+                  ") and outputs (",
+                  ctx.getNumOutputs(),
+                  ")");
+            }
+            int64_t totalDim = 0;
+            for (int64_t d : split) {
+              totalDim += d;
+            }
+            if (totalDim != splitDimValue) {
+              fail_shape_inference(
+                  "Mismatch between the sum of 'split' (",
+                  totalDim,
+                  ") and the split dimension of the input (",
+                  splitDimValue,
+                  ")");
+            }
+          } else {
+            int numOutputs = static_cast<int>(ctx.getNumOutputs());
+            if (splitDimValue % numOutputs != 0) {
+              fail_shape_inference("The input is not evenly splittable");
+            }
+            int chunkSize = splitDimValue / numOutputs;
+            for (int i = 0; i < static_cast<int>(ctx.getNumOutputs()); i++) {
+              split.push_back(chunkSize);
```

Comment:
nit 1: It seems a little ugly to store the same value 'ChunkSize' in the vector for each output. it would be great if the single value 'chunkSize' can just be handled below when you actually assign the dim value.

nit 2: 'chunkSize' doesn't seem c++ style. You can consider using 'chunk_size'

Reply:
As for 2, fixed.

As for 1, let me double-check if you like the change, since I feel current code is better. If I follow your suggestion, I think we'll have two copies of the for loop which assigns `dim_value` like

```
if (getRepeatedAttribute(ctx, "split", split)) {
   ...
   for (size_t i = 0; i < ctx.getNumOutputs(); i++) {
       *ctx.getOutputType(i)->mutable_tensor_type()->mutable_shape() =
            shape;
        ctx.getOutputType(i)
            ->mutable_tensor_type()
             ->mutable_shape()
             ->mutable_dim(axis)
             ->set_dim_value(split[i]);
    }
} else {
  ...
  int chunk_size = split_dim_value / num_outputs;
   for (size_t i = 0; i < ctx.getNumOutputs(); i++) {
       *ctx.getOutputType(i)->mutable_tensor_type()->mutable_shape() =
            shape;
        ctx.getOutputType(i)
            ->mutable_tensor_type()
             ->mutable_shape()
             ->mutable_dim(axis)
             ->set_dim_value(chunk_size);
    }
}
```
Is this what you want?

Reply:
That's fine. We can just keep it the way it is for now.